java/eks/private-cluster/src/main/java/com/amazonaws/cdk/examples/EksPrivateClusterStack.java

```diff
@@ -0,0 +1,228 @@
+                        .build()))
+            .subnetSelection(
+                SubnetSelection.builder().subnetType(SubnetType.PRIVATE_ISOLATED).build())
+            .securityGroup(this.cluster.getClusterSecurityGroup())
+            .instanceType(InstanceType.of(InstanceClass.T3, InstanceSize.SMALL))
+            .build();
+
+    // allow the client role to assume the cluster admin role
+    clusterAdmin.grantAssumeRole(client.getRole());
+
+    // add permission to allow aws eks update-kubeconfig i.e. DescribeCluster
+    client
+        .getInstance()
+        .addToRolePolicy(
+            PolicyStatement.Builder.create()
+                .effect(Effect.ALLOW)
+                .actions(List.of("eks:DescribeCluster", "eks:ListClusters"))
+                .resources(List.of("*"))
+                .build());
+
+    // NOTE: In production, restrict the access to specific repositories and buckets
+    // add ECR access for client to pull images
+    // use ECR pull through to get public images like nginx
+    client
+        .getRole()
+        .addManagedPolicy(
+            ManagedPolicy.fromAwsManagedPolicyName("AmazonEC2ContainerRegistryReadOnly"));
+    // access to read assets from S3 bucket e.g. kubectl, awscliv2, etc
+    client
+        .getRole()
+        .addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName("AmazonS3ReadOnlyAccess"));
+
+    // add user data to install kubectl and apply the update-kubeconfig
+    // add in the env variable
+    // export AWS_STS_REGIONAL_ENDPOINTS=regional
+  }
+
+  private void addEndpoints() {
+    List<InterfaceVpcEndpointAwsService> endpoints =
+        List.of(
+            InterfaceVpcEndpointAwsService.ECR,
+            InterfaceVpcEndpointAwsService.ECR_DOCKER,
+            InterfaceVpcEndpointAwsService.CLOUDWATCH_MONITORING,
+            InterfaceVpcEndpointAwsService.CLOUDWATCH_LOGS,
+            InterfaceVpcEndpointAwsService.EVENTBRIDGE,
+            InterfaceVpcEndpointAwsService.STS,
+            InterfaceVpcEndpointAwsService.SSM,
+            InterfaceVpcEndpointAwsService.SSM_MESSAGES,
+            InterfaceVpcEndpointAwsService.LAMBDA,
+            InterfaceVpcEndpointAwsService.EKS,
+            InterfaceVpcEndpointAwsService.EC2,
+            InterfaceVpcEndpointAwsService.EC2_MESSAGES,
+            InterfaceVpcEndpointAwsService.CLOUDFORMATION,
+            InterfaceVpcEndpointAwsService.STEP_FUNCTIONS,
+            InterfaceVpcEndpointAwsService.STEP_FUNCTIONS_SYNC);
+
+    SubnetSelection subnets =
+        SubnetSelection.builder().subnetType(SubnetType.PRIVATE_ISOLATED).build();
+
+    for (InterfaceVpcEndpointAwsService e : endpoints) {
+      vpc.addInterfaceEndpoint(
+          e.getShortName(),
+          InterfaceVpcEndpointOptions.builder().service(e).subnets(subnets).build());
+    }
+
+    vpc.addGatewayEndpoint(
+        "s3", GatewayVpcEndpointOptions.builder().service(GatewayVpcEndpointAwsService.S3).build());
+  }
```

Comment:
WDYT of advising the developer (through a documentation or comment) that this will considerably increase the costs of networking and suggest adding only the endpoints that make sense for their infrastructure?

Reply:
Thanks for the feedback. I have updated the README with a note on the VPC Endpoint pricing. After further testing, have removed CloudFormation and EventBridge in the endpoints as they are not needed.